File: apps/researchtree/memory/spec.py
"""Intent and spec documents: repo config, sections, includes and section-level changes.

Python port of apps/core/src/spec.ts; both must pass tests/fixtures/spec-cases.json.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Sequence

from ruamel.yaml import YAML
from ruamel.yaml.error import YAMLError

REPO_CONFIG_PATH = ".researchtree.yml"
DEFAULT_SPEC_PATH = "SPEC.md"
DEFAULT_INTENT_PATH = "INTENT.md"
SECTION_LINE_BUDGET = 40
_MAX_INCLUDE_DEPTH = 5

_FENCE_RE = re.compile(r"^\s{0,3}(`{3,}|~{3,})")
_HEADING_RE = re.compile(r"^\s{0,3}(#{1,6})(?:[ \t]+(.*?))?[ \t]*$")
_ID_RE = re.compile(r"<!--\s*id:\s*([A-Za-z0-9._-]+)\s*-->")
_ID_LINE_RE = re.compile(r"^\s*<!--\s*id:\s*([A-Za-z0-9._-]+)\s*-->\s*$")
_INCLUDE_RE = re.compile(r"^\s*<!--\s*include:\s*(\S+?)\s*-->\s*$")
_PREFIX_RE = re.compile(r"^(?!/)(?!.*//)(?!.*\.\.)[A-Za-z0-9._\-/]+/$")
_SCHEME_RE = re.compile(r"^[a-z][a-z0-9+.-]*:", re.IGNORECASE)


@dataclass(frozen=True)
class Section:
    key: str
    id: str | None
    level: int
    title: str
    summary: str | None
    body: str
    parent: str | None

    def to_dict(self) -> dict[str, Any]:
        return {
            "key": self.key,
            "id": self.id,
            "level": self.level,
            "title": self.title,
            "summary": self.summary,
            "body": self.body,
            "parent": self.parent,
        }


@dataclass(frozen=True)
class SectionChange:
    key: str
    kind: str  # added | removed | changed | renamed
    title: str
    before: Section | None
    after: Section | None

    def __str__(self) -> str:
        return f"{self.kind}: {self.title or '(머리말)'}"


@dataclass
class Spec:
    """An assembled spec document at one git ref."""

    path: str
    title: str | None
    text: str
    sections: list[Section]
    missing: list[str] = field(default_factory=list)
    ref: str | None = None

    def __getitem__(self, key: str) -> Section:
        """By key (`vocoder/sample-rate`) or exact title."""
        for s in self.sections:
            if s.key == key:
                return s
        for s in self.sections:
            if s.title == key:
                return s
        raise KeyError(f"no section {key!r}")

    def __iter__(self):
        return iter(self.sections)

    def __len__(self) -> int:
        return len(self.sections)

    def summary(self) -> str:
        return spec_summary(self.sections)

    def diff(self, before: Spec | None) -> list[SectionChange]:
        """Changes from `before` to this spec."""
        return diff_specs(before.sections if before else None, self.sections)

    def check(self) -> list[dict[str, Any]]:
        return check_spec(self.sections, self.missing)

    def __repr__(self) -> str:
        return f"Spec({self.path!r}, {len(self.sections)} sections)"


class _Fences:
    def __init__(self) -> None:
        self.open: str | None = None

    def step(self, line: str) -> bool:
        m = _FENCE_RE.match(line)
        if self.open is None:
            if m:
                self.open = m.group(1)
                return True
            return False
        if m and m.group(1)[0] == self.open[0] and len(m.group(1)) >= len(self.open) and line.strip() == m.group(1):
            self.open = None
        return True


def slugify(title: str) -> str:
    """Lowercase, and every run of characters other than letters and digits becomes one `-`."""
    return re.sub(r"[\W_]+", "-", title.lower()).strip("-")


def _trim_block(lines: list[str]) -> str:
    out = [line.rstrip() for line in lines]
    while out and out[0] == "":
        out.pop(0)
    while out and out[-1] == "":
        out.pop()
    return "\n".join(out)


def _summary_of(body: str) -> str | None:
    lines = body.split("\n")
    first = next((i for i, line in enumerate(lines) if line.strip()), None)
    if first is None or not lines[first].lstrip().startswith(">"):
        return None
    quote: list[str] = []
    for line in lines[first:]:
        s = line.lstrip()
        if not s.startswith(">"):
            break
        quote.append(re.sub(r"^>\s?", "", s).strip())
    text = " ".join(q for q in quote if q).strip()
    return text or None


def _split_lines(text: str) -> list[str]:
    return re.split(r"\r?\n", text)


def parse_spec(text: str) -> list[Section]:
    """Split Markdown into sections, one per ATX heading (headings inside code fences do not count)."""
    raws: list[dict[str, Any]] = [{"level": 0, "title": "", "id": None, "lines": [], "fresh": False}]
    fences = _Fences()
    for line in _split_lines(text):
        cur = raws[-1]
        if fences.step(line):
            cur["lines"].append(line)
            cur["fresh"] = False
            continue
        m = _HEADING_RE.match(line)
        if m:
            title = re.sub(r"[ \t]+#+$", "", m.group(2) or "")
            title = re.sub(r"^#+$", "", title)
            mid = _ID_RE.search(title)
            if mid:
                title = _ID_RE.sub("", title)
            raws.append({"level": len(m.group(1)), "title": title.strip(), "id": mid.group(1) if mid else None, "lines": [], "fresh": True})
            continue
        id_line = _ID_LINE_RE.match(line) if cur["fresh"] and cur["id"] is None else None
        if id_line:
            cur["id"] = id_line.group(1)
        else:
            cur["lines"].append(line)
        cur["fresh"] = False

    sections: list[Section] = []
    seen: set[str] = set()
    stack: list[Section] = []
    for r in raws:
        body = _trim_block(r["lines"])
        if r["level"] == 0:
            if body:
                sections.append(Section("", None, 0, "", _summary_of(body), body, None))
            continue
        while stack and stack[-1].level >= r["level"]:
            stack.pop()
        parent = stack[-1] if stack else None
        # The document title (level 1) is left out of child keys, so renaming it keeps every section's history.
        prefix = f"{parent.key}/" if parent and parent.level >= 2 else ""
        key = r["id"] or f"{prefix}{slugify(r['title']) or 'section'}"
        if key in seen:
            n = 2
            while f"{key}-{n}" in seen:
                n += 1
            key = f"{key}-{n}"
        seen.add(key)
        section = Section(key, r["id"], r["level"], r["title"], _summary_of(body), body, parent.key if parent else None)
        sections.append(section)
        stack.append(section)
    return sections


def join_path(dir: str, rel: str) -> str | None:
    """Resolve `rel` against `dir` inside the repository; None when it leaves it or is not a plain path."""
    if not rel or "\\" in rel or _SCHEME_RE.match(rel):
        return None
    parts = [] if rel.startswith("/") else [p for p in dir.split("/") if p]
    for seg in rel.split("/"):
        if seg in ("", "."):
            continue
        if seg == "..":
            if not parts:
                return None
            parts.pop()
        else:
            parts.append(seg)
    return "/".join(parts) if parts else None


def _dirname(path: str) -> str:
    i = path.rfind("/")
    return "" if i < 0 else path[:i]


def load_spec(entry: str, read: Callable[[str], str | None]) -> Spec | None:
    """Read the entry file and expand `<!-- include: path -->` lines. None when the entry file does not exist."""
    missing: list[str] = []

    def expand(path: str, stack: list[str]) -> str | None:
        text = read(path)
        if text is None:
            return None
        out: list[str] = []
        fences = _Fences()
        for line in _split_lines(text):
            m = None if fences.step(line) else _INCLUDE_RE.match(line)
            if not m:
                out.append(line)
                continue
            target = join_path(_dirname(path), m.group(1))
            sub = expand(target, [*stack, target]) if target and target not in stack and len(stack) < _MAX_INCLUDE_DEPTH else None
            if sub is None:
                missing.append(target or m.group(1))
                out.append(line)
            else:
                out.append(sub[:-1] if sub.endswith("\n") else sub)
        return "\n".join(out)

    text = expand(entry, [entry])
    if text is None:
        return None
    sections = parse_spec(text)
    title = next((s.title for s in sections if s.level == 1), None)
    return Spec(entry, title, text, sections, missing)


def diff_specs(before: Sequence[Section] | None, after: Sequence[Section]) -> list[SectionChange]:
    """Section-level changes from `before` to `after` (None before = everything is new)."""
    old = {s.key: s for s in before or []}
    now = {s.key for s in after}
    changes: list[SectionChange] = []
    for a in after:
        b = old.get(a.key)
        if b is None:
            changes.append(SectionChange(a.key, "added", a.title, None, a))
        elif b.body != a.body:
            changes.append(SectionChange(a.key, "changed", a.title, b, a))
        elif b.title != a.title:
            changes.append(SectionChange(a.key, "renamed", a.title, b, a))
    for b in before or []:
        if b.key not in now:
            changes.append(SectionChange(b.key, "removed", b.title, b, None))
    return changes


def section_history(versions: Iterable[tuple[str, Sequence[Section] | None]], key: str) -> list[dict[str, str]]:
    """In which versions (oldest first) a section appeared, changed, was renamed or was removed."""
    out: list[dict[str, str]] = []
    prev: Sequence[Section] | None = None
    for name, sections in versions:
        cur = sections or []
        change = next((c for c in diff_specs(prev, cur) if c.key == key), None)
        if change:
            out.append({"version": name, "kind": change.kind})
        prev = cur
    return out


def check_spec(sections: Sequence[Section], missing: Sequence[str] = ()) -> list[dict[str, Any]]:
    """Readability checks: unresolved includes, sections without a summary line, sections over the line budget."""
    issues: list[dict[str, Any]] = [{"kind": "missing-include", "path": p} for p in missing]
    for s in sections:
        if s.level >= 2 and s.summary is None:
            issues.append({"kind": "no-summary", "key": s.key})
        lines = len(s.body.split("\n")) if s.body else 0
        if lines > SECTION_LINE_BUDGET:
            issues.append({"kind": "too-long", "key": s.key, "lines": lines})
    return issues


def spec_summary(sections: Sequence[Section]) -> str:
    """Headings and their summary lines only: the whole design on one page."""
    out = []
    for s in sections:
        if s.level == 0:
            continue
        pad = "  " * max(0, s.level - 1)
        out.append(f"{pad}{s.title} — {s.summary}" if s.summary else f"{pad}{s.title}")
    return "\n".join(out)


def parse_repo_config(text: str) -> tuple[dict[str, str], list[dict[str, str]]]:
    """Parse `.researchtree.yml`. Unknown keys are reported but otherwise ignored; bad values are dropped."""
    config: dict[str, str] = {}
    warnings: list[dict[str, str]] = []
    try:
        data = YAML(typ="safe").load(text)
    except YAMLError:
        return config, [{"code": "invalid-yaml"}]
    if data is None:
        return config, warnings
    if not isinstance(data, dict):
        return config, [{"code": "not-mapping"}]
    for key, value in data.items():
        if key == "prefix":
            p = value.strip() if isinstance(value, str) else ""
            if p and not p.endswith("/"):
                p += "/"
            if _PREFIX_RE.match(p):
                config["prefix"] = p
            else:
                warnings.append({"code": "invalid-prefix", "key": "prefix"})
        elif key in ("spec", "intent"):
            path = join_path("", value.strip()) if isinstance(value, str) else None
            if path:
                config[key] = path
            else:
                warnings.append({"code": "invalid-path", "key": key})
        else:
            warnings.append({"code": "unknown-key", "key": str(key)})
    return config, warnings

File: apps/core/src/spec.ts
import { isMap, parseDocument } from "yaml";

/**
 * Intent and spec documents: the repo config file, splitting a spec into sections, assembling
 * included files, and section-level changes between two versions of a spec.
 * The Python port is apps/researchtree/memory/spec.py; both must pass tests/fixtures/spec-cases.json.
 */

/** Repository settings shared by the whole team, read from the root branch. */
export const REPO_CONFIG_PATH = ".researchtree";
/** What the file used to be called. Repositories that still use it are read the same way. */
export const REPO_CONFIG_PATH_OLD = ".researchtree.yml";
/** Where to look for the repository's settings, in the order they win. */
export const REPO_CONFIG_PATHS = [REPO_CONFIG_PATH, REPO_CONFIG_PATH_OLD] as const;
export const DEFAULT_SPEC_PATH = "SPEC.md";
export const DEFAULT_INTENT_PATH = "INTENT.md";

/** A section longer than this (own lines, without children) is flagged by checkSpec. */
export const SECTION_LINE_BUDGET = 40;
const MAX_INCLUDE_DEPTH = 5;

export interface RepoConfig {
  /** Root branch of the research tree. The file sits on the default branch, so it can name it. */
  root?: string;
  prefix?: string;
  spec?: string;
  intent?: string;
}

export type RepoConfigWarning =
  | { code: "invalid-yaml" }
  | { code: "not-mapping" }
  | { code: "invalid-branch"; key: "root" | "prefix" }
  | { code: "invalid-path"; key: "spec" | "intent" }
  | { code: "unknown-key"; key: string };

export interface SpecSection {
  /**
   * Stable identity: the explicit id, else the parent's key + "/" + the title's slug (a level-1
   * parent is left out). "" for text before the first heading.
   */
  key: string;
  /** Set with `<!-- id: name -->` on the heading line or the line right after it. */
  id: string | null;
  /** 1-6 for headings, 0 for text before the first heading. */
  level: number;
  title: string;
  /** Text of the `>` quote that opens the section, if any. */
  summary: string | null;
  /** The section's own text (up to the next heading of any level), without the id comment. */
  body: string;
  parent: string | null;
}

export interface Spec {
  /** Entry file the spec was assembled from. */
  path: string;
  /** Title of the first level-1 heading. */
  title: string | null;
  /** Assembled Markdown (includes expanded). */
  text: string;
  sections: SpecSection[];
  /** Included paths that could not be read (or were not allowed). */
  missing: string[];
}

export type SectionChangeKind = "added" | "removed" | "changed" | "renamed";

export interface SectionChange {
  key: string;
  kind: SectionChangeKind;
  /** Title after the change (before, for a removed section). */
  title: string;
  before: SpecSection | null;
  after: SpecSection | null;
}

export type SpecIssue =
  | { kind: "missing-include"; path: string }
  | { kind: "no-summary"; key: string }
  | { kind: "too-long"; key: string; lines: number };

const FENCE_RE = /^\s{0,3}(`{3,}|~{3,})/;
const HEADING_RE = /^\s{0,3}(#{1,6})(?:[ \t]+(.*?))?[ \t]*$/;
const ID_RE = /<!--\s*id:\s*([A-Za-z0-9._-]+)\s*-->/;
const ID_LINE_RE = /^\s*<!--\s*id:\s*([A-Za-z0-9._-]+)\s*-->\s*$/;
const INCLUDE_RE = /^\s*<!--\s*include:\s*(\S+?)\s*-->\s*$/;

/** Tracks fenced code blocks so headings and comments inside them are ignored. */
class Fences {
  private open: string | null = null;

  /** Returns true when the line is inside a fence (fence lines themselves included). */
  step(line: string): boolean {
    const m = FENCE_RE.exec(line);
    if (this.open === null) {
      if (m) {
        this.open = m[1]!;
        return true;
      }
      return false;
    }
    if (m && m[1]![0] === this.open[0] && m[1]!.length >= this.open.length && line.trim() === m[1]) this.open = null;
    return true;
  }
}

/** Lowercase, and every run of characters other than letters and digits becomes one `-`. */
export function slugify(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^\p{L}\p{N}]+/gu, "-")
    .replace(/^-+|-+$/g, "");
}

function trimBlock(lines: string[]): string {
  const out = lines.map((l) => l.replace(/\s+$/, ""));
  while (out.length && out[0] === "") out.shift();
  while (out.length && out[out.length - 1] === "") out.pop();
  return out.join("\n");
}

function summaryOf(body: string): string | null {
  const lines = body.split("\n");
  const first = lines.findIndex((l) => l.trim() !== "");
  if (first < 0 || !lines[first]!.trimStart().startsWith(">")) return null;
  const quote: string[] = [];
  for (const l of lines.slice(first)) {
    const s = l.trimStart();
    if (!s.startsWith(">")) break;
    quote.push(s.replace(/^>\s?/, "").trim());
  }
  const text = quote.filter(Boolean).join(" ").trim();
  return text || null;
}

/** Split Markdown into sections, one per ATX heading (headings inside code fences do not count). */
export function parseSpec(text: string): SpecSection[] {
  type Raw = { level: number; title: string; id: string | null; lines: string[]; fresh: boolean };
  const raws: Raw[] = [{ level: 0, title: "", id: null, lines: [], fresh: false }];
  const fences = new Fences();
  for (const line of text.split(/\r?\n/)) {
    const cur = raws[raws.length - 1]!;
    if (fences.step(line)) {
      cur.lines.push(line);
      cur.fresh = false;
      continue;
    }
    const m = HEADING_RE.exec(line);
    if (m) {
      let title = (m[2] ?? "").replace(/[ \t]+#+$/, "").replace(/^#+$/, "");
      const id = ID_RE.exec(title);
      if (id) title = title.replace(ID_RE, "");
      raws.push({ level: m[1]!.length, title: title.trim(), id: id ? id[1]! : null, lines: [], fresh: true });
      continue;
    }
    const idLine = cur.fresh && cur.id === null ? ID_LINE_RE.exec(line) : null;
    if (idLine) cur.id = idLine[1]!;
    else cur.lines.push(line);
    cur.fresh = false;
  }

  const sections: SpecSection[] = [];
  const seen = new Set<string>();
  const stack: SpecSection[] = [];
  for (const r of raws) {
    const body = trimBlock(r.lines);
    if (r.level === 0) {
      if (body) sections.push({ key: "", id: null, level: 0, title: "", summary: summaryOf(body), body, parent: null });
      continue;
    }
    while (stack.length && stack[stack.length - 1]!.level >= r.level) stack.pop();
    const parent = stack[stack.length - 1] ?? null;
    // The document title (level 1) is left out of child keys, so renaming it keeps every section's history.
    const prefix = parent && parent.level >= 2 ? `${parent.key}/` : "";
    let key = r.id ?? `${prefix}${slugify(r.title) || "section"}`;
    if (seen.has(key)) {
      let n = 2;
      while (seen.has(`${key}-${n}`)) n++;
      key = `${key}-${n}`;
    }
    seen.add(key);
    const section: SpecSection = { key, id: r.id, level: r.level, title: r.title, summary: summaryOf(body), body, parent: parent?.key ?? null };
    sections.push(section);
    stack.push(section);
  }
  return sections;
}

/** Resolve `rel` against directory `dir` inside the repository. Null when it leaves the repository or is not a plain path. */
export function joinPath(dir: string, rel: string): string | null {
  if (!rel || rel.includes("\\") || /^[a-z][a-z0-9+.-]*:/i.test(rel)) return null;
  const parts = rel.startsWith("/") ? [] : dir.split("/").filter(Boolean);
  for (const seg of rel.split("/")) {
    if (seg === "" || seg === ".") continue;
    if (seg === "..") {
      if (!parts.length) return null;
      parts.pop();
    } else parts.push(seg);
  }
  return parts.length ? parts.join("/") : null;
}

function dirname(path: string): string {
  const i = path.lastIndexOf("/");
  return i < 0 ? "" : path.slice(0, i);
}

/**
 * Read the entry file and expand `<!-- include: path -->` lines (relative to the including file,
 * outside code fences, at most five levels deep). Null when the entry file does not exist.
 */
export async function loadSpec(entry: string, read: (path: string) => Promise<string | null>): Promise<Spec | null> {
  const missing: string[] = [];
  const expand = async (path: string, stack: string[]): Promise<string | null> => {
    const text = await read(path);
    if (text === null) return null;
    const out: string[] = [];
    const fences = new Fences();
    for (const line of text.split(/\r?\n/)) {
      const m = fences.step(line) ? null : INCLUDE_RE.exec(line);
      if (!m) {
        out.push(line);
        continue;
      }
      const target = joinPath(dirname(path), m[1]!);
      const sub = target && !stack.includes(target) && stack.length < MAX_INCLUDE_DEPTH ? await expand(target, [...stack, target]) : null;
      if (sub === null) {
        missing.push(target ?? m[1]!);
        out.push(line);
      } else out.push(sub.replace(/\n$/, ""));
    }
    return out.join("\n");
  };
  const text = await expand(entry, [entry]);
  if (text === null) return null;
  const sections = parseSpec(text);
  return { path: entry, title: sections.find((s) => s.level === 1)?.title ?? null, text, sections, missing };
}

/** Section-level changes from `before` to `after` (null before = everything is new). Order: after's order, then removed sections. */
export function diffSpecs(before: readonly SpecSection[] | null, after: readonly SpecSection[]): SectionChange[] {
  const old = new Map((before ?? []).map((s) => [s.key, s]));
  const now = new Set(after.map((s) => s.key));
  const changes: SectionChange[] = [];
  for (const a of after) {
    const b = old.get(a.key);
    if (!b) changes.push({ key: a.key, kind: "added", title: a.title, before: null, after: a });
    else if (b.body !== a.body) changes.push({ key: a.key, kind: "changed", title: a.title, before: b, after: a });
    else if (b.title !== a.title) changes.push({ key: a.key, kind: "renamed", title: a.title, before: b, after: a });
  }
  for (const b of before ?? []) {
    if (!now.has(b.key)) changes.push({ key: b.key, kind: "removed", title: b.title, before: b, after: null });
  }
  return changes;
}

/** In which versions (oldest first) a section appeared, changed, was renamed or was removed. */
export function sectionHistory(
  versions: readonly { name: string; sections: readonly SpecSection[] | null }[],
  key: string,
): { version: string; kind: SectionChangeKind }[] {
  const out: { version: string; kind: SectionChangeKind }[] = [];
  let prev: readonly SpecSection[] | null = null;
  for (const v of versions) {
    const cur = v.sections ?? [];
    const change = diffSpecs(prev, cur).find((c) => c.key === key);
    if (change) out.push({ version: v.name, kind: change.kind });
    prev = cur;
  }
  return out;
}

/** Readability checks: unresolved includes, sections without a summary line, sections over the line budget. */
export function checkSpec(spec: Pick<Spec, "sections" | "missing">): SpecIssue[] {
  const issues: SpecIssue[] = spec.missing.map((path) => ({ kind: "missing-include", path }));
  for (const s of spec.sections) {
    if (s.level >= 2 && s.summary === null) issues.push({ kind: "no-summary", key: s.key });
    const lines = s.body ? s.body.split("\n").length : 0;
    if (lines > SECTION_LINE_BUDGET) issues.push({ kind: "too-long", key: s.key, lines });
  }
  return issues;
}

/** Headings and their summary lines only: the whole design on one page. */
export function specSummary(spec: Pick<Spec, "sections">): string {
  const out: string[] = [];
  for (const s of spec.sections) {
    if (s.level === 0) continue;
    const pad = "  ".repeat(Math.max(0, s.level - 1));
    out.push(s.summary ? `${pad}${s.title} — ${s.summary}` : `${pad}${s.title}`);
  }
  return out.join("\n");
}

const PREFIX_RE = /^(?!\/)(?!.*\/\/)(?!.*\.\.)[A-Za-z0-9._\-/]+\/$/;
const BRANCH_RE = /^(?!\/)(?!.*\/\/)(?!.*\.\.)(?!.*\/$)[A-Za-z0-9._\-/]+$/;

/** Parse `.researchtree`. Unknown keys are reported but otherwise ignored; bad values are dropped. */
export function parseRepoConfig(text: string): { config: RepoConfig; warnings: RepoConfigWarning[] } {
  const warnings: RepoConfigWarning[] = [];
  const config: RepoConfig = {};
  const doc = parseDocument(text);
  if (doc.errors.length) return { config, warnings: [{ code: "invalid-yaml" }] };
  const data: unknown = doc.toJS();
  if (data === null || data === undefined) return { config, warnings };
  if (!isMap(doc.contents) || typeof data !== "object" || Array.isArray(data)) return { config, warnings: [{ code: "not-mapping" }] };
  for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
    if (key === "root") {
      const root = typeof value === "string" ? value.trim() : "";
      if (root && BRANCH_RE.test(root)) config.root = root;
      else warnings.push({ code: "invalid-branch", key: "root" });
    } else if (key === "prefix") {
      let p = typeof value === "string" ? value.trim() : "";
      if (p && !p.endsWith("/")) p += "/";
      if (PREFIX_RE.test(p)) config.prefix = p;
      else warnings.push({ code: "invalid-branch", key: "prefix" });
    } else if (key === "spec" || key === "intent") {
      const path = typeof value === "string" ? joinPath("", value.trim()) : null;
      if (path) config[key] = path;
      else warnings.push({ code: "invalid-path", key });
    } else warnings.push({ code: "unknown-key", key });
  }
  if (config.root && config.prefix && config.root.startsWith(config.prefix)) {
    delete config.root;
    warnings.push({ code: "invalid-branch", key: "root" });
  }
  return { config, warnings };
}

/** Line-level diff for showing one section's change. Falls back to delete-all / add-all for very long text. */
export function lineDiff(before: string, after: string): { op: " " | "+" | "-"; text: string }[] {
  const a = before ? before.split("\n") : [];
  const b = after ? after.split("\n") : [];
  if (a.length * b.length > 250_000) return [...a.map((text) => ({ op: "-" as const, text })), ...b.map((text) => ({ op: "+" as const, text }))];
  const n = a.length;
  const m = b.length;
  // lcs[i][j] = length of the longest common subsequence of a[i:] and b[j:]
  const lcs: number[][] = Array.from({ length: n + 1 }, () => new Array<number>(m + 1).fill(0));
  for (let i = n - 1; i >= 0; i--) {
    for (let j = m - 1; j >= 0; j--) lcs[i]![j] = a[i] === b[j] ? lcs[i + 1]![j + 1]! + 1 : Math.max(lcs[i + 1]![j]!, lcs[i]![j + 1]!);
  }
  const out: { op: " " | "+" | "-"; text: string }[] = [];
  let i = 0;
  let j = 0;
  while (i < n && j < m) {
    if (a[i] === b[j]) {
      out.push({ op: " ", text: a[i]! });
      i++;
      j++;
    } else if (lcs[i + 1]![j]! >= lcs[i]![j + 1]!) out.push({ op: "-", text: a[i++]! });
    else out.push({ op: "+", text: b[j++]! });
  }
  while (i < n) out.push({ op: "-", text: a[i++]! });
  while (j < m) out.push({ op: "+", text: b[j++]! });
  return out;
}
